### PR TITLE
Revert "Revert "[Job Submission][refactor 1/N] Add AgentInfo to GCSNodeInfo (…"

### PR DESCRIPTION
**Pull request by Catch-Bull**
Reverts ray-project/ray#27242

fix flaky test: test_ray_shutdown.py

After the raylet dies, the CoreWorker will have two ways to exit:
1. Finished the construction of CPP class CoreWorker, it will exit by monitor thread.
2. Before the construction of CPP class CoreWorker, it will raise Exception when trying to connect GCS

the worst-case in **case 2**: it will get time out when [`check_version_info`](https://github.com/ray-project/ray/blob/729566d8ffb1b5e4ac6f8b8ef76cc1a445a33c64/python/ray/_private/worker.py#L1900), then [try to publish error to GCS](https://github.com/ray-project/ray/blob/729566d8ffb1b5e4ac6f8b8ef76cc1a445a33c64/python/ray/_private/worker.py#L1906), it will get timeout too, but publish timeout is 60s, our default timeout of function `wait_for_condition` is 10s, so the test case will fail.

### CODE REVIEW

**Comment by Catch-Bull** [comment]
@rkooo567 I've been testing on my machine for over an hour and should be fine, I'll retry UT multiple times to make sure it's no longer flaky.

**Comment by Catch-Bull** [comment]
@rkooo567 HI, I retry UT multiple times, and I think it works!
image
image
the failed one, which causes by a existing flaky test case:
image

File: python/ray/_private/worker.py
Comment (SongGuyang) [inline python/ray/_private/worker.py:1905]: Add a comment here.

**Review by SongGuyang — APPROVED** [review]

File: python/ray/_private/worker.py
Comment (Catch-Bull) [inline python/ray/_private/worker.py:1909]: if `node.check_version_info` raise grpc.RpcError, it means GCS is unreachable, so the `ray._private.utils.publish_error_to_driver` will fail too, and we raise the Exception early can avoid wait to publish timeout, it will cost 60s for nothing

**Review by rkooo567 — CHANGES_REQUESTED** [review]
Please do not merge it yet. I will look at it soon.

File: python/ray/tests/BUILD
Comment (Catch-Bull) [inline python/ray/tests/BUILD:297]: the current PR will make `cluster.add_node` takes 1 second more than before, There are nearly 30 `cluster.add_node` in `test_placement_group_2.py`, it takes about 240s before this PR, So I think it is reasonable to increase the time limit of `test_placement_group_2` to avoid flaky tests.

**Comment by Catch-Bull** [comment]
@rkooo567 Hi, could you review this PR soon?

File: src/ray/protobuf/common.proto
Comment (rkooo567) [inline src/ray/protobuf/common.proto:693]: newline!

File: python/ray/tests/BUILD
Comment (rkooo567) [reply python/ray/tests/BUILD:297]: > cluster.add_node takes 1 second more than before

This actually sounds pretty bad... (I feel like it will increase the runtime of tests too long).

Why is this the case?

**Review by rkooo567 — COMMENTED** [review]
Mostly LGTM, and the fix makes sense. After we resolve the below thing, we can merge! 

1. I understand the fix for the issue, but why does it happen after this PR? The PR seems unrelated to me.
2. There are total 519 `add_node` (and I guess if we include for loop, it could easily exceed 1000), meaning the test runtime will increase 10~20 minutes. Maybe this much of overhead is acceptable because our test runtime is pretty long, but I still would like to understand why `add_node` takes 1 more second?

**Comment by Catch-Bull** [comment]
> Mostly LGTM, and the fix makes sense. After we resolve the below thing, we can merge!
> 
> 1. I understand the fix for the issue, but why does it happen after this PR? The PR seems unrelated to me.
> 2. There are total 519 `add_node` (and I guess if we include for loop, it could easily exceed 1000), meaning the test runtime will increase 10~20 minutes. Maybe this much of overhead is acceptable because our test runtime is pretty long, but I still would like to understand why `add_node` takes 1 more second?

@rkooo567 Hi:
1. the driver just needs to call `notify_raylet` to ensure the raylet is ready (this does not need the raylet finished registered to GCS), but the scheduling of normal tasks needs to make sure the raylet finished registered. 
   - **In general:** the time gap between driver start and the worker process start(this means raylet finished register) will take 1 more second than before. the time of kill driver is almost equal, when the raylet/GCS is killed, the worker process has a greater chance of not completing check_version_info
2. The reason why `add_node` takes 1 more second is raylet needs to wait for the agent finished registered before registering itself with GCS, before this PR, the raylet register itself to GCS, and the agent register itself to raylet is async. and the default value of `add_node`'s argument `wait` is `True`, I can set it to `False` and call `wait_for_nodes` in the end of cluster creation.

**Comment by Catch-Bull** [comment]
@rkooo567 about why I think the raylet register itself to GCS will take 1 more second:
without this PR, the time gap between raylet sending the RegisterRequest and HanldeAgentRegister is one second.

File: src/ray/protobuf/common.proto
Comment (Catch-Bull) [reply src/ray/protobuf/common.proto]: done.

**Comment by rkooo567** [comment]
> the time gap between raylet sending the RegisterRequest and HanldeAgentRegister is one second.

Hmm this seems too long? Can you actually measure it (maybe log how long it takes)?

**Review by rkooo567 — APPROVED** [review]

**Comment by Catch-Bull** [comment]
> > the time gap between raylet sending the RegisterRequest and HanldeAgentRegister is one second.
> 
> Hmm this seems too long? Can you actually measure it (maybe log how long it takes to register agent)?

@rkooo567 In current branch:
[raylet][14:10:15,059]  Raylet launch the agent process.
[raylet][14:10:15,060]  the first time to call `Raylet::RegisterGcs().` 
[agent][14:10:15,843] DashboardAgent.__init__.
[agent][14:10:16,160] agent send register request to raylet. 
[raylet][14:10:16,167]  agent finished registering.
[raylet][14:10:16,167]  send register request to GCS.

The main time overhead is in the initialization of the agent process.

**Comment by rkooo567** [comment]
I see. So the most of overhead is from the process start up

**Comment by rkooo567** [comment]
I will merge this PR when it is ready

**Comment by iycheng** [comment]
LGTM!

**Review by fishbone — APPROVED** [review]

**Comment by SongGuyang** [comment]
@Catch-Bull @rkooo567 Ready to merge?

**Comment by SongGuyang** [comment]
The failed tune tests are flakey in master.

**Pull request merged**
